src/doc/tarpl/vec-alloc.md

```diff
@@ -0,0 +1,117 @@
+% Allocating Memory
+
+So:
+
+```rust,ignore
+#![feature(heap_api)]
+
+use std::rt::heap::EMPTY;
+use std::mem;
+
+impl<T> Vec<T> {
+    fn new() -> Self {
+        assert!(mem::size_of::<T>() != 0, "We're not ready to handle ZSTs");
+        unsafe {
+            // need to cast EMPTY to the actual ptr type we want, let
+            // inference handle it.
+            Vec { ptr: Unique::new(heap::EMPTY as *mut _), len: 0, cap: 0 }
+        }
+    }
+}
+```
+
+I slipped in that assert there because zero-sized types will require some
+special handling throughout our code, and I want to defer the issue for now.
+Without this assert, some of our early drafts will do some Very Bad Things.
+
+Next we need to figure out what to actually do when we *do* want space. For that,
+we'll need to use the rest of the heap APIs. These basically allow us to
+talk directly to Rust's instance of jemalloc.
+
+We'll also need a way to handle out-of-memory conditions. The standard library
+calls the `abort` intrinsic, but calling intrinsics from normal Rust code is a
+pretty bad idea. Unfortunately, the `abort` exposed by the standard library
+allocates. Not something we want to do during `oom`! Instead, we'll call
+`std::process::exit`.
+
+```rust
+fn oom() {
+    ::std::process::exit(-9999);
+}
+```
+
+Okay, now we can write growing. Roughly, we want to have this logic:
+
+```text
+if cap == 0:
+    allocate()
+    cap = 1
+else
+    reallocate
+    cap *= 2
+```
+
+But Rust's only supported allocator API is so low level that we'll need to
+do a fair bit of extra work, though. We also need to guard against some special
+conditions that can occur with really large allocations. In particular, we index
+into arrays using unsigned integers, but `ptr::offset` takes signed integers. This
+means Bad Things will happen if we ever manage to grow to contain more than
+`isize::MAX` elements. Thankfully, this isn't something we need to worry about
+in most cases.
+
+On 64-bit targets we're artifically limited to only 48-bits, so we'll run out
```

Comment:
The 48-bit limit is true on current x86-64 chips (at the hardware level) and it looks like it may be true on AArch64 too (unsure how strict/hardware this limit is), but it's not a fundamental property of all 64-bit targets, and certainly not a limitation Rust will impose on itself.